src/hotspot/share/opto/addnode.cpp

```diff
@@ -395,9 +395,159 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
+  // power-of-2 addition (e.g., 3 * a => (a << 2) + a). Without this check, GVN would keep trying to optimize the same
+  // node and can't progress. For example, 3 * a => (a << 2) + a => 3 * a => (a << 2) + a => ...
+  if (find_power_of_two_addition_pattern(this, bt, nullptr) != nullptr) {
+    return nullptr;
+  }
+
+  Node* in1 = in(1);
+  Node* in2 = in(2);
+  jlong multiplier;
+
+  // While multiplications can be potentially optimized to power-of-2 subtractions (e.g., a * 7 => (a << 3) - a),
+  // (x - y) + y => x is already handled by the Identity() methods. So, we don't need to check for that pattern here.
+  if (find_simple_addition_pattern(in1, bt, &multiplier) == in2
+      || find_simple_lshift_pattern(in1, bt, &multiplier) == in2
+      || find_simple_multiplication_pattern(in1, bt, &multiplier) == in2
+      || find_power_of_two_addition_pattern(in1, bt, &multiplier) == in2) {
+    multiplier++; // +1 for the in2 term
+
+    Node* con = (bt == T_INT)
+                ? (Node*) phase->intcon((jint) multiplier) // intentional type narrowing to allow overflow at max_jint
+                : (Node*) phase->longcon(multiplier);
+    return MulNode::make(con, in2, bt);
+  }
+
+  return nullptr;
+}
+
+// Try to match `a + a`. On success, return `a` and set `2` as `multiplier`.
+// The method matches `n` for pattern: AddNode(a, a).
+Node* AddNode::find_simple_addition_pattern(Node* n, BasicType bt, jlong* multiplier) {
+  if (n->Opcode() == Op_Add(bt) && n->in(1) == n->in(2)) {
+    *multiplier = 2;
+    return n->in(1);
+  }
+
+  return nullptr;
+}
+
+// Try to match `a << CON`. On success, return `a` and set `1 << CON` as `multiplier`.
+// Match `n` for pattern: LShiftNode(a, CON).
+// Note that the power-of-2 multiplication optimization could potentially convert a MulNode to this pattern.
+Node* AddNode::find_simple_lshift_pattern(Node* n, BasicType bt, jlong* multiplier) {
+  // Note that power-of-2 multiplication optimization could potentially convert a MulNode to this pattern
+  if (n->Opcode() == Op_LShift(bt) && n->in(2)->is_Con()) {
+    Node* con = n->in(2);
+    if (con->is_top()) {
+      return nullptr;
+    }
+
+    *multiplier = java_shift_left(1, con->get_int(), bt);
+    return n->in(1);
+  }
+
+  return nullptr;
+}
+
+// Try to match `CON * a`. On success, return `a` and set `CON` as `multiplier`.
+// Match `n` for patterns:
+//     - MulNode(CON, a)
+//     - MulNode(a, CON)
+Node* AddNode::find_simple_multiplication_pattern(Node* n, BasicType bt, jlong* multiplier) {
+  // This optimization technically only produces MulNode(CON, a), but we might as match MulNode(a, CON), too.
+  if (n->Opcode() == Op_Mul(bt) && (n->in(1)->is_Con() || n->in(2)->is_Con())) {
+    Node* con = n->in(1);
+    Node* base = n->in(2);
+
+    // swap ConNode to lhs for easier matching
+    if (!con->is_Con()) {
+      swap(con, base);
+    }
```

Comment:
Is that necessary? Does `Mul` not automatically get canonicalized so that the constant is on the rhs?

Reply:
This is not related to `Mul` canonicalization. Swapping those two variables makes my next 4 lines syntactically easier to write. (So I don't have to do `(con->is_Con ? con : base)->is_top()` and so on.) This shouldn't be any more costly with a modern C++ compiler.

Reply:
Obsolete after refactored to use `find_simple_lshift_pattern`.